docs/knowledgebase/runtime/macros.md

```diff
@@ -189,6 +202,217 @@ also inserts lifecycle trait implementations for `Module`, e.g.
+
+### #[pallet::hooks]
+
+**When to Use**
+
+Required for runtime hooks.
+
+**What it Does**
+
+In `decl_module` we used to declare hooks logic by implementing [each method](https://crates.parity.io/frame_support/traits/trait.Hooks.html#provided-methods). In FRAME V2, `Hooks` are made available by using the `Hooks` trait allowing developers to implement each individual trait.
+
+For example: 
+
+```rust
+#[pallet::hooks]
+imple<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+  // Hooks functions and logic goes here.
+}
+```
+
+**Docs**
+- See the [documentation](https://crates.parity.io/frame_support/attr.pallet.html#hooks-pallethooks-mandatory)
+- See the [macro expansion](https://substrate.dev/rustdocs/v3.0.0/frame_support/attr.pallet.html#macro-expansion-2)
+
+### #[pallet::call]
+
+**When to Use**
+
+Required, to implement a pallet's dispatchables. Each dispatchable must:
+- define a weight with the #[pallet::weight($expr)] attribute
+- have its first argument as `origin: OriginFor<T>`, 
+- use compact encoding for argument using #[pallet::compact]
+- return `DispatchResultWithPostInfo` or `DispatchResult`
+
+**What it Does**
+
+Similar to `decl_module`, it allows to create dispatchable functions which will generate associated items from the `impl` code blocks. In other words, it aggregates all dispatchable logic using the [`Call` enum](https://substrate.dev/rustdocs/v3.0.0/frame_system/pallet/enum.Call.html) which will aggregate all dispatchable calls into a single runtime call.
+
+**Docs**
+- See the [documentation](https://crates.parity.io/frame_support/attr.pallet.html#call-palletcall-mandatory)
+
+### #[pallet::error]
+
+**When to Use**
+
+Optionally, to define errors types from the pallet's dispatchables.
+
+**What it Does**
+
+Puts error types into metadata.
+
+**Docs**
+- See [documentation](https://crates.parity.io/frame_support/attr.pallet.html#error-palleterror-optional)
+
+### #[pallet::events]
+
+**When to Use**
+
+Optionally, to define a pallet's event types. 
+
+
+**What it Does**
+
+It is similar to errors but it holds more types. They generate the metadata of the event and add_derive. It uses the `#[pallet::metadata(..)]` attribute to define what metadata to put from the events.
+
+For example: 
+```rust
+#[pallet::event]
+#[pallet::metadata(u32 = "SpecialU32")]
+pub enum Event<T: Config> {
+	Proposed(u32, T::AccountId),
+}
+```
+
+**Docs**
+- See [documentation](https://substrate.dev/rustdocs/v3.0.0/frame_support/attr.pallet.html#event-palletevent-optional)
+### #[pallet::storage]
+
+**When to Use** 
+
+Optionally, as it can be used multiple times.
+
+**What it Does** 
+
+To define some abstract storage inside runtime storage.
+
+**Note:** this changes how storage is declared compared to v1. Here, `[pallet::storage]` is using Rust like storage maps to do storage, such that it can be generic over the pallet. 
```

Comment:
> > is using Rust like storage maps
> 
> I'm not sure I understand, do you mean it is using rust in a similar way as HashMap type in the standard library.

Yes, this was a comparison you made in Seminar, I'm not sure I understood correctly.
> > such that it can be generic over the pallet
>
Same. Here I understand this to mean: we can use it multiple times. 

How would you describe how _"this changes how storage is declared compared to v1"_?

Reply:
Ah ok if the comparison with HashMap in std is:
* the type HashMap is a generic struct which takes Key, Value, and Hasher generically.
* the type StorageMap follow the same principle, it is a struct which takes Key, Value, Hasher, and also Prefix (and also some other parameter in order to change the behavior a bit: QueryKind and OnEmpty).
The Prefix generic defines the place where the storage will store its `(key, value)` in the storage (i.e. the trie).
The Prefix generic defines the pallet prefix and the storage prefix, storagemap will use to store its key after the prefix: `concat(twox_128(pallet_prefix), towx_128(storage_prefix))` which is unique for the storage.
The macro will replace `_` with a generated Prefix implementation, the generated implementation for Prefix use the pallet name as pallet prefix and the storage name as storage prefix.

(the pallet name is the name given to the pallet in construct_runtime, the storage name is the ident: (e.g. "Foo" in `type Foo<T> = ...`))